src/instructlab/model/full_train.py

```diff
@@ -0,0 +1,319 @@
+    # Third Party
+    from instructlab.training import config
+    from instructlab.training import data_process as dp
+    from instructlab.training import (
+        multipack_sampler,
+        token_dataset,
+        tokenizer_utils,
+        utils,
+    )
+    from torch.utils.data import DataLoader
+    import torch
+
+    dp.main(
+        config.DataProcessArgs(
+            data_output_path=train_args.data_output_dir,
+            model_path=train_args.model_path,
+            data_path=train_args.data_path,
+            max_seq_len=train_args.max_seq_len,
+            chat_tmpl_path=train_args.chat_tmpl_path,
+        )
+    )
+
+    # load chat template based on path in the args
+    CHAT_TEMPLATE, SPECIAL_TOKENS = utils.retrieve_chat_template(
+        train_args.chat_tmpl_path
+    )
+    # get the tokenizer for the model
+    tokenizer = tokenizer_utils.setup_tokenizer(
+        train_args.model_path, SPECIAL_TOKENS, CHAT_TEMPLATE
+    )
+
+    # setup the dataset and place it in data.jsonl, this needs to be used for training NOT the jsonl produced by sdg
+    dataset = token_dataset.setup_dataset(
+        os.path.join(train_args.data_output_dir, "data.jsonl"),
+    )
+
+    # based on the length of the dataset, figure out the max batch len
+    packing_max_batch_len, _ = (
+        multipack_sampler.find_packing_max_batch_len_and_grad_accum(
+            num_gpus=1,
+            avg_sample_len=dataset.get_lengths().mean(),
+            effective_batch_size=train_args.effective_batch_size,
+            max_batch_len_per_gpu=train_args.max_batch_len,
+            is_padding=train_args.is_padding_free,
+            pad_id=tokenizer.pad_token_id,
+            dataset=dataset,
+            seed=47,
+        )
+    )
+
+    collate_fn = partial(pad_collate_fn, pad_token_id=tokenizer.pad_token_id)
+
+    # use a multipack sampler since this is a non-dist. scenario, training library defaults to this as well
+    sampler = multipack_sampler.MultipackDistributedBatchSampler(
+        batch_max_length=packing_max_batch_len,
+        lengths=dataset.get_lengths(),
+        num_replicas=1,
+        rank=0,
+        seed=47,
+        padding=True,
+    )
+    sampler = {"batch_sampler": sampler}
+
+    # 4 workers for the dataloader as compared to the original 8 to optimize performance
+    # this dataloader needs a spawn multiproc method to optimize memory and to work on Apple Silicon.
+    dataloader = DataLoader(
+        dataset,
+        **sampler,
+        num_workers=4,
+        collate_fn=collate_fn,
+    )
+    dataloader.multiprocessing_context = "spawn"
+
+    # set device based off argument given
+    dev = torch.device(device)
+    # auto model based on model path
+    model = AutoModelForCausalLM.from_pretrained(train_args.model_path)
+    if len(tokenizer) > model.config.vocab_size:
+        logger.info(
+            f"WARNING: tokenizer has {len(tokenizer)} tokens but model has {model.config.vocab_size} vocab size"
+        )
```

Comment:
Using `logger.info` and having `WARNING` at the beginning isn't good practice - use `logger.warning` instead